src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -941,5 +960,232 @@ export async function getV2CandidatePools({
+export async function getMixedRouteCandidatePools({
+  tokenIn,
+  tokenOut,
+  routeType,
+  routingConfig,
+  V3subgraphProvider,
+  V2subgraphProvider,
+  tokenProvider,
+  V3poolProvider,
+  V2poolProvider,
+  blockedTokenListProvider,
+  chainId,
+}: MixedRouteGetCandidatePoolsParams): Promise<{
+  V2poolAccessor: V2PoolAccessor;
+  V3poolAccessor: V3PoolAccessor;
+  candidatePools: CandidatePoolsBySelectionCriteria;
+  subgraphPools: (V2SubgraphPool | V3SubgraphPool)[];
+}> {
+  const { blockNumber } = routingConfig;
+  const { subgraphPools: V3subgraphPools, candidatePools: V3candidatePools } =
+    await getV3CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V3poolProvider,
+      routeType,
+      subgraphProvider: V3subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+  const { subgraphPools: V2subgraphPools, candidatePools: V2candidatePools } =
+    await getV2CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V2poolProvider,
+      routeType,
+      subgraphProvider: V2subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+
+  /**
+   * Main heuristic for pruning mixedRoutes:
+   * - we pick V2 pools with higher liq than respective V3 pools, or if the v3 pool doesn't exist
+   *
+   * This way we can reduce calls to our provider since it's possible to generate a lot of mixed routes
+   */
+  /// We only really care about pools involving the tokenIn or tokenOut explictly,
+  /// since there's no way a long tail token in V2 would be routed through as an intermediary
+  const V2topByTVLPools = [
+    ...V2candidatePools.selections.topByTVLUsingTokenIn,
+    ...V2candidatePools.selections.topByBaseWithTokenIn,
+    /// tokenOut:
+    ...V2candidatePools.selections.topByTVLUsingTokenOut,
+    ...V2candidatePools.selections.topByBaseWithTokenOut,
+    /// Direct swap:
+    ...V2candidatePools.selections.topByDirectSwapPool,
+  ];
+
+  const V2topByTVLSortedPools = _(V2subgraphPools)
+    .filter((pool) => V2topByTVLPools.map((p) => p.id).includes(pool.id))
+    .sortBy((pool) => -pool.reserveUSD)
+    .value();
+
+  /// we consider all returned V3 pools for this heuristic to "fill in the gaps"
+  const V3sortedPools = _(V3subgraphPools)
+    .sortBy((pool) => -pool.tvlUSD)
+    .value();
+
+  /// Finding pools with greater reserveUSD on v2 than tvlUSD on v3, or if there is no v3 liquidity
+  const buildV2Pools: V2SubgraphPool[] = [];
+  V2topByTVLSortedPools.map((V2subgraphPool) => {
+    const V3subgraphPool = V3sortedPools.find(
+      (pool) =>
+        (pool.token0.id == V2subgraphPool.token0.id &&
+          pool.token1.id == V2subgraphPool.token1.id) ||
+        (pool.token0.id == V2subgraphPool.token1.id &&
+          pool.token1.id == V2subgraphPool.token0.id)
+    );
+
+    if (V3subgraphPool) {
+      if (V2subgraphPool.reserveUSD > V3subgraphPool.tvlUSD) {
```

Comment:
No action here, but just a comment: ideally we would want to compare V2 reserve USD to V3 TVL within +/-% of the current price on V3. Concentrated liquidity makes comparisons like this super weird as all the liquidity could be out of range and thus not being touched during the swap, so TVL can make a V3 pool look better than it really is.

I think this is the best we can do currently and should give good results so no action needed though!

Reply:
Yeah I used to have it be something like 1.5x more TVL on V2 than V3 but we were missing alot of good pools. Can consider this something to do in the future!